flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java

```diff
@@ -0,0 +1,163 @@
+ *     if the production is done. The actual release can take some time
+ *     if 'the end of consumption' confirmation is being awaited implicitly
+ *     or the partition is later released by {@code releasePartitions(Collection<ResultPartitionID>)}.</li>
+ *     <li>{@code releasePartitions(Collection<ResultPartitionID>)} is called outside of the task thread,
+ *     e.g. to manage the local resource lifecycle of external partitions which outlive the task production.</li>
+ * </ol>
+ * The partitions, which currently still occupy local resources, can be queried with {@code getUnreleasedPartitions}.
+ *
+ * <h3>Input gate management.</h3>
+ *
+ * <p>The interface implements a factory for the task input gates: {@code createInputGates}.
+ * The created gates are grouped per task and handed over to the task thread upon its startup.
+ * The task is responsible for the gates' lifecycle from that moment.
+ *
+ * <p>When tha task is deployed and the input gates are created, it can happen that not all consumed partitions
+ * are known at that moment e.g. because their producers have not been started yet.
+ * Therefore, the {@link ShuffleEnvironment} provides a method {@code updatePartitionInfo} to update them
+ * externally, ouside of the task thread, when the producer becomes known.
+ */
+public interface ShuffleEnvironment {
+
+	/**
+	 * Start the internal related services upon {@link TaskExecutor}'s startup.
+	 *
+	 * @return a port to connect to the task executor for shuffle data exchange, -1 if only local connection is possible.
+	 */
+	int start() throws IOException;
+
+	/**
+	 * Shutdown the internal related services upon {@link TaskExecutor}'s stop.
+	 */
+	void shutdown();
+
+	/**
+	 * Factory method for the task's {@link ResultPartitionWriter}s.
+	 *
+	 * @param taskName the task name, used for logs
+	 * @param executionAttemptID execution attempt id of the task
+	 * @param resultPartitionDeploymentDescriptors descriptors of the partition, produced by the task
+	 * @param outputGroup shuffle specific group for output metrics
+	 * @param buffersGroup shuffle specific group for buffer metrics
+	 * @return array of the task's {@link ResultPartitionWriter}s
+	 */
+	ResultPartitionWriter[] createResultPartitionWriters(
+		String taskName,
+		ExecutionAttemptID executionAttemptID,
+		Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors,
+		MetricGroup outputGroup,
+		MetricGroup buffersGroup);
+
+	/**
+	 * Release local resources occupied with the given partitions.
+	 *
+	 * @param partitionIds partition ids to release
+	 */
+	void releasePartitions(Collection<ResultPartitionID> partitionIds);
+
+	/**
+	 * Report unreleased partitions.
+	 *
+	 * @return collection of partitions which still occupy some resources locally on this task executor
+	 * and have been not released yet.
+	 */
+	Collection<ResultPartitionID> getUnreleasedPartitions();
```

Comment:
Maybe better to call this method `getLocalPartitions` or `getPartitionsOccupyingLocalResources`. In the case of an external shuffle service, this method would return an empty collection even though some partitions might still be unreleased but stored somewhere else.